src/wiseagents/wise_agent_impl.py

```diff
@@ -632,6 +635,312 @@ def set_response_delivery(self, response_delivery: Callable[[], WiseAgentMessage
+            description (str): a description of the agent
+            transport (WiseAgentTransport): the transport to use for communication
+            llm (WiseAgentLLM): the LLM to use for coordinating the collaboration
+            phases (Optional[List[str]]): the optional list of phases, defaults to "Data Collection" and "Data Analysis"
+            max_iterations (Optional[int]): the maximum number of iterations to run the phases, defaults to 5
+            confidence_score_threshold (Optional[int]): the confidence score threshold to determine if the final answer
+            is acceptable, defaults to 85
+        """
+        self._name = name
+        self._route_response_to = ""
+        self._phases = phases if phases is not None else ["Data Collection", "Data Analysis"]
+        self._max_iterations = max_iterations
+        self._confidence_score_threshold = confidence_score_threshold
+        super().__init__(name=name, description=description, transport=transport, llm=llm)
+
+    def __repr__(self):
+        """Return a string representation of the agent."""
+        return (f"{self.__class__.__name__}(name={self.name}, description={self.description}, transport={self.transport},"
+                f"llm={self.llm}, phases={self.phases},max_iterations={self.max_iterations}")
+
+    @property
+    def phases(self) -> List[str]:
+        """Get the list of phases."""
+        return self._phases
+
+    @property
+    def max_iterations(self) -> int:
+        """Get the maximum number of iterations."""
+        return self._max_iterations
+
+    @property
+    def confidence_score_threshold(self) -> int:
+        """Get the confidence score threshold."""
+        return self._confidence_score_threshold
+
+    def process_request(self, request):
+        """
+        Process a request message by kicking off the collaboration in phases.
+
+        Args:
+            request (WiseAgentMessage): the request message to process
+        """
+        logging.debug(f"Coordinator received request: {request}")
+        self._route_response_to = request.sender
+
+        # Generate a chat ID that will be used to collaborate on this query
+        chat_id = str(uuid.uuid4())
+
+        ctx = WiseAgentRegistry.get_or_create_context(request.context_name)
+
+        # Determine the agents required to answer the query
+        agent_selection_prompt = ("Given the following query and a description of the agents that are available," +
+                                  " determine all of the agents that could be required to solve the query." +
+                                  " Format the response as a space separated list of agent names and don't include " +
+                                  " anything else in the response.\n" +
+                                  " Query: " + request.message + "\n" + "Available agents:\n" +
+                                  "\n".join(WiseAgentRegistry.get_agent_names_and_descriptions()) + "\n")
```

Comment:
Note that with this prompt, it's likely that the response will only contain a subset of the agents required to solve the problem. However, the next prompt below for assigning agents to phases identifies the agents that were missed from this prompt. We might be able to remove this prompt or modify it to use sub-task decomposition to try to identify all the required agents.